### ISSUE TITLE
The code problem of Custom Feature Extractor 

### ISSUE THREAD

**Issue by aahuhu** [DLR-RM/stable-baselines3#794]
**Important Note: We do not do technical support, nor consulting** and don't answer personal questions per email.
Please post your question on the [RL Discord](https://discord.com/invite/xhfNqQv), [Reddit](https://www.reddit.com/r/reinforcementlearning/) or [Stack Overflow](https://stackoverflow.com/) in that case.

### 📚 Documentation

####problem
In [this code](https://github.com/DLR-RM/stable-baselines3/blob/master/docs/guide/custom_policy.rst#custom-feature-extractor)
1:what's the function of  PART"with  torch.no_grad...." ? why use torch.no_grad to compute "n_flatten"
2: what's the meaning of n_flatten ?

### Checklist

- [x] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/) (**required**)
- [x] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)

###

**Comment by araffin** [comment]
Hello,
i highly recommend you to read more about convolutional neural networks (cnn) and pytorch first.

The `no_grad()` is to save computation time and memory because we don't need gradients.
We want to know the shape of the output of the cnn, so we can deduce its dimension when flattened (`n_flatten`) in order to use fully connected layers afterward.

**Comment by goweiting** [comment]
Hello @araffin,

I stumbedupon this comment while searching for solution to a similar issue. It is now solved. 

The BaseFeatureExtractor uses attribute features_dim to build the first layer of mlp defined in net_arch. This is however not clear in the documentation. I only manage to understand what it was meant in the documentation after toying with the numbers and observing the changes in the `model.policy` output.

```
:param features_dim: (int) Number of features extracted.
        This corresponds to the number of unit for the last layer.
```
in https://stable-baselines3.readthedocs.io/en/master/guide/custom_policy.html?highlight=basefeature 


My recommendation is to make clear that features_dim in BaseFeatureExtractor, shall be equivalent to the output, instead of enforcing the addition of adding another linear layer at the end of the extraction. Hope i make sense here. For your consideration.

**Comment by araffin** [comment]
> My recommendation is to make clear that features_dim in BaseFeatureExtractor, shall be equivalent to the output, instead of enforcing the addition of adding another linear layer at the end of the extraction

what do you mean? could you elaborate?
Adding a linear layer at the end of the CNN is a common practice to reduce dimension (and computation).

**Comment by goweiting** [comment]
Hey, sorry i should be clearer.
Referencing the feature_extractor interfacing in the subsequent net_arch as described [here](https://stable-baselines3.readthedocs.io/en/master/guide/custom_policy.html?highlight=feature%20extractor#custom-feature-extractor), i found that the net_arch infer the size of the input by referencing the feature_extractor's attribute - features_dim. 

I took some time to understand this, because I was using a squeezing MLPolicy of dense layers (300 (input), 256, 128, 64, 32) while defining features_dim as 300 (thinking it was the input features's dim). However, sbl3 was throwing error that the net_arch policy was expecting a size of 300. 

This is why i think the argument (features_dim) that is being passed in Feature Extractor was confusing, as i would have thought it refers to the dimensions of features, as compared to what was expected - the dimensions of the linear layer before net_arch is "attached".

**Comment by araffin** [comment]
> as i would have thought it refers to the dimensions of features, as compared to what was expected - the dimensions of the linear layer before net_arch is "attached".

You mean dimension of the observation? (we would have call it observation in that case ;))

maybe that diagram in the doc would have helped you: https://stable-baselines3.readthedocs.io/en/master/guide/custom_policy.html

**Comment by goweiting** [comment]
haha! we were referencing the same diagram. it was helpful,.... 

My interpretation on first sight is features_dim could be referring to the dimension of the input to the feature extractor or the output of the extractor (which is what it was intended to be). 

Nevertheless, it is not a big issue.

**Comment by araffin** [comment]
closing as original question was answered then ;)
